read-fonts/src/tables/postscript/scale.rs

```diff
@@ -0,0 +1,541 @@
+    /// to retrieve the subfont index for the desired glyph and then
+    /// [`subfont`](Self::subfont) to create an instance of the subfont for a
+    /// particular size and location in variation space.
+    /// Creating subfont instances is not free, so this process is exposed in
+    /// discrete steps to allow for caching.
+    ///
+    /// The result is emitted to the specified pen.
+    pub fn outline(
+        &self,
+        subfont: &ScalerSubfont,
+        glyph_id: GlyphId,
+        coords: &[F2Dot14],
+        pen: &mut impl Pen,
+    ) -> Result<(), Error> {
+        use super::charstring;
+        let charstring_data = self
+            .top_dict
+            .charstrings
+            .as_ref()
+            .ok_or(Error::MissingCharstrings)?
+            .get(glyph_id.to_u16() as usize)?;
+        let subrs = subfont.subrs(self)?;
+        let blend_state = subfont.blend_state(self, coords)?;
+        let mut pen_sink = charstring::PenSink::new(pen);
+        let mut simplifying_adapter = charstring::SimplifyingSink::new(&mut pen_sink);
+        let scale = if subfont.size <= 0.0 {
+            Fixed::ONE
+        } else {
+            // Note: we do an intermediate scale to 26.6 to ensure we
+            // match FreeType
+            Fixed::from_bits((subfont.size * 64.) as i32)
+                / Fixed::from_bits(self.units_per_em as i32)
+        };
+        let mut scaling_adapter =
+            charstring::ScalingSink26Dot6::new(&mut simplifying_adapter, scale);
+        // TODO: hinting will be another sink adapter that slots in here
+        charstring::evaluate(
+            charstring_data,
+            self.global_subrs(),
+            subrs,
+            blend_state,
+            &mut scaling_adapter,
+        )?;
+        simplifying_adapter.finish();
+        Ok(())
+    }
+
+    fn offset_data(&self) -> FontData<'a> {
+        match &self.version {
+            Version::Version1(cff1) => cff1.offset_data(),
+            Version::Version2(cff2) => cff2.offset_data(),
+        }
+    }
+
+    fn global_subrs(&self) -> Index<'a> {
+        match &self.version {
+            Version::Version1(cff1) => cff1.global_subrs().into(),
+            Version::Version2(cff2) => cff2.global_subrs().into(),
+        }
+    }
+
+    fn private_dict_range(&self, subfont_index: u32) -> Result<Range<usize>, Error> {
+        if let Some(font_dicts) = &self.top_dict.font_dicts {
+            // If we have a font dict index, use that
+            let font_dict_data = font_dicts.get(subfont_index as usize)?;
+            let mut range = None;
+            for entry in dict::entries(font_dict_data, None) {
+                if let dict::Entry::PrivateDictRange(r) = entry? {
+                    range = Some(r);
+                    break;
+                }
+            }
+            range
+        } else {
+            // Otherwise, assume the top dict provided a private dict range
```

Comment:
"assume" threw me here; it seems to not quite be true, we error if not ok below. I think perhaps it's more like `// Last chance, a privatedict range from the top dict` than "assume"?